src/test/blockmanager_tests.cpp

```diff
@@ -82,4 +83,113 @@ BOOST_FIXTURE_TEST_CASE(blockmanager_scan_unlink_already_pruned_files, TestChain
+    CChain chain{};
+
+    // Test blocks with no transactions, not even a coinbase
+    CBlock block1;
+    block1.nVersion = 1;
+    CBlock block2;
+    block2.nVersion = 2;
+    CBlock block3;
+    block3.nVersion = 3;
+
+    // They are 80 bytes header + 1 byte 0x00 for vtx length
+    int TEST_BLOCK_SIZE = 81;
+
+    // Blockstore is empty
+    BOOST_CHECK_EQUAL(blockman.CalculateCurrentUsage(), 0);
+
+    // Write first block
+    FlatFilePos pos1 = blockman.SaveBlockToDisk(block1, 1, chain, *params, nullptr);
+
+    // Write second block
+    FlatFilePos pos2 = blockman.SaveBlockToDisk(block2, 2, chain, *params, nullptr);
+
+    // Two blocks in the file
+    BOOST_CHECK_EQUAL(blockman.CalculateCurrentUsage(), (TEST_BLOCK_SIZE + BLOCK_SERIALIZATION_HEADER_SIZE) * 2);
+
+    // First two blocks are written as expected
+    CBlock read_block;
+    node::ReadBlockFromDisk(read_block, pos1, params->GetConsensus());
+    BOOST_CHECK_EQUAL(read_block.nVersion, 1);
+    node::ReadBlockFromDisk(read_block, pos2, params->GetConsensus());
+    BOOST_CHECK_EQUAL(read_block.nVersion, 2);
+
+    // Record last modified time of blk00000.dat
+    const fs::path file_path = node::GetBlockPosFilename(pos2);
+    const fs::file_time_type time1 = fs::last_write_time(file_path);
+
+#ifdef WIN32
+    // Ensure that if the file is modified its timestamp is updated,
+    // even on Windows (FAT has 2 second resolution)
+    std::this_thread::sleep_for(std::chrono::milliseconds(2000));
+#endif
+
+    // Attempt to save block 3 to original position of block 2
+    blockman.SaveBlockToDisk(block3, 3, chain, *params, &pos2);
+
+    // Still only two blocks in the file
+    BOOST_CHECK_EQUAL(blockman.CalculateCurrentUsage(), (TEST_BLOCK_SIZE + BLOCK_SERIALIZATION_HEADER_SIZE) * 2);
+
+    // Block 2 was not overwritten:
+    //   SaveBlockToDisk() did not call WriteBlockToDisk() because `FlatFilePos* dbp` was non-null
+    node::ReadBlockFromDisk(read_block, pos2, params->GetConsensus());
+    BOOST_CHECK_EQUAL(read_block.nVersion, 2);
+
+    // blk00000.dat was not modified
+    const fs::file_time_type time2 = fs::last_write_time(file_path);
+    BOOST_CHECK(time1 == time2);
+
+    // Chain requires a tip to continue (for undo data handling in FindBlockPos())
+    CBlockIndex tip;
+    tip.nHeight = 1;
+    chain.SetTip(tip);
+
+#ifdef WIN32
+    std::this_thread::sleep_for(std::chrono::milliseconds(2000));
+#endif
+
+    // Now fill up the first file to fastprune limit and start new block file
+    CBlock dummy_block;
+    FlatFilePos last_block;
+    while (last_block.nFile < 1) {
+        last_block = blockman.SaveBlockToDisk(dummy_block, 0, chain, *params, nullptr);
+    }
+
+    // Record last modified time of blk00000.dat now that it is full and finished
+    const fs::file_time_type time3 = fs::last_write_time(file_path);
+
+    // Sanity check: blk00000.dat should have been modified writing the dummy blocks.
+    BOOST_CHECK(time2 != time3);
```

Comment:
This seems more problematic than what I mentioned above because this can cause false-positive. Is the resolution one second? Could all that happened since `time2` was sampled be in the same second? But I ran the test many times and did not see the problem.

Reply:
What I learned is that windows/FAT systems have 2-second resolution but Linux/EXT4 etc have ms or ns resolution. I had a hard time determining the best method for testing when a file is flushed but I think you're right, using timestamps is just asking for intermittency. I think I wanted to avoid hashing files because it means opening the file while blockmanager might already have it open... but I think that's going to be the best way to test.

Reply:
ok trying another approach where we ALWAYS pause, to ensure incremental timestamps. But on windows we have to pause for 2 seconds. All other filesystems we only need to wait for 1 ms.